File: modulefederation/libs/portal/src/lib/version/version-range.ts
import { Version } from "./version"

type VersionComparator = (version: Version) => boolean

export class VersionRange {
    // instance data

    private operations: VersionComparator[] = []

    // constructor

    constructor(version: string) {
        this.parse(version)
    }

    // private

    private compare(version: Version, op: string) : VersionComparator{
        switch (op) {
            case "<":
                return (v: Version) => v.lt(version)
            case "<=":
                return (v: Version) => v.le(version)
            case ">":
                return (v: Version) => v.gt(version)
            case ">=":
                return (v: Version) => v.ge(version)
            default:
                throw new Error("bad operator")
        }
    }

    private parse(version: string) {
        const exp = /^(?<op>(>|>=|<|<=))(?<v>\d[.\d]+)(,(?<op1>(>|>=|<|<=))(?<v1>\d[.\d]+))*$/
        const result = exp.exec(version)
        if ( result?.groups ) {
            this.operations.push(this.compare(new Version(result.groups["v"]), result.groups["op"]))

            if ( result?.groups["op1"])
                this.operations.push(this.compare(new Version(result.groups["v1"]), result.groups["op1"]))
        }
        else throw new Error(`could not parse version '${version}'`)
    }

    // public

    matches(version: Version) : boolean {
        for ( const operation of this.operations)
            if ( !operation(version))
                return false

        return true
    }
}

File: portal/core/src/main/java/com/serious/portal/version/VersionRange.kt
package com.serious.portal.version
/*
 * @COPYRIGHT (C) 2023 Andreas Ernst
 *
 * All rights reserved
 */

import java.util.regex.Matcher
import java.util.regex.Pattern

class VersionRange(version: String) {
        // instance data

        private var operations: ArrayList<(Version) -> Boolean > = ArrayList()

        // constructor

        init {
            parse(version)
        }

        // private

        private fun compare(version: Version, op: String) : (Version) -> Boolean {
            when (op) {
                "<" -> return { v -> v.lt(version) }
                "<=" -> return { v -> v.le(version) }
                ">" -> return { v -> v.gt(version) }
                ">=" -> return { v -> v.ge(version) }

                else -> {
                    throw Error("bad operator")
                }
            }
        }

        private fun parse(version: String) {
            val pattern = "(?<op>(>|>=|<|<=))(?<v>\\d[.\\d]+)(,(?<op1>(>|>=|<|<=))(?<v1>\\d[.\\d]+))*"
            val regex: Pattern = Pattern.compile(pattern)
            val matcher: Matcher = regex.matcher(version)
            val success: Boolean = matcher.find()

            if ( success ) {
                val op = matcher.group("op")
                val v  = matcher.group("v")

                operations.add(compare(Version(matcher.group("v")), matcher.group("op")))

                if ( matcher.group("op1") != null)
                    operations.add(compare(Version(matcher.group("v1")), matcher.group("op1")))

            }
            else throw Error("parse error")
        }

        // public

        fun matches(version: Version) : Boolean {
            for ( operation in operations)
                if ( !operation(version))
                    return false

            return true
        }
}